peps/pep-0749.rst

```diff
@@ -0,0 +1,741 @@
+:py:mod:`dataclasses` and :py:class:`typing.NamedTuple`), so it makes sense to include
+it in the standard library.
+
+*Add this functionality to a private module*: It would be possible to initially develop
+the module in a private standard library module (e.g., ``_annotations``), and publicize
+it only after we have gained more experience with the API. However, we already know
+that we will need parts of this module for the standard library itself (e.g., for
+implementing :py:mod:`!dataclasses` and :py:class:`!typing.NamedTuple`). Even if we make
+it private, the module will inevitably get used by third-party users. It is preferable
+to start with a clear, documented API from the beginning, to enable third-party users to
+support :pep:`649` semantics as thoroughly as the standard library. The module will
+immediately be used in other parts of the standard library, ensuring that it covers a
+reasonable set of use cases.
+
+Behavior of the REPL
+====================
+
+:pep:`649` specifies the following behavior of the interactive REPL:
+
+    For the sake of simplicity, in this case we forego delayed evaluation.
+    Module-level annotations in the REPL shell will continue to work exactly
+    as they do with “stock semantics”, evaluating immediately and setting the
+    result directly inside the ``__annotations__`` dict.
+
+There are several problems with this proposed behavior. It makes the REPL the
+only context where annotations are still evaluated immediately, which is
+confusing for users and complicates the language.
+
+It also makes the implementation of the REPL more complex, as it needs to
+ensure that all statements are compiled in "interactive" mode, even if their
+output does not need to be displayed. (This matters if there are multiple
+statements in a single line evaluated by the REPL.)
+
+Most importantly, this breaks some plausible use cases that inexperienced
+users could run into. A user might write the following in a file::
+
+    x: X | None = None
+    class X: ...
+
+Under :pep:`649` this would work fine: ``X`` is not yet defined when it is used,
+but it is not evaluated. However, if a user were to paste this same code into the
+REPL and execute it line by line, it would throw a ``NameError``, because the
+name ``X`` is not yet defined.
+
+This topic was previously discussed `on Discourse <https://discuss.python.org/t/pep-649-behavior-of-the-repl/54109>`__.
+
+Specification
+-------------
+
+We propose to treat the interactive console like any other module-level code, and
+make annotations lazily evaluated. This makes the language more consistent and
+avoids subtle behavior changes between modules and the REPL.
+
+Because the REPL is evaluated line by line, we would generate a new ``__annotate__``
+function for every evaluated line that contains annotations. Whenever a line
+containing annotations is evaluated, the previous ``__annotate__`` function is
+lost:
```

Comment:
Does this only apply in the REPL's "module namespace"? Or are e.g. annotations in class namespaces also affected?

Reply:
Just the global namespace. I'll add a paragraph to clarify.